src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStream.cs

```diff
@@ -114,40 +119,60 @@ public override Task FlushAsync(CancellationToken cancellationToken)
             return WriteAsync(null, 0, 0, cancellationToken);
         }
 
-        private async ValueTask<int> ReadAsyncInternal(Memory<byte> destination, CancellationToken cancellationToken)
+        private ValueTask<int> ReadAsyncInternal(Memory<byte> destination, CancellationToken cancellationToken)
         {
-            while (true)
+            try
             {
-                var result = await _input.ReadAsync(cancellationToken);
-                var readableBuffer = result.Buffer;
-                try
+                while (true)
                 {
-                    if (_throwOnCancelled && result.IsCanceled && _cancelCalled)
+                    if (!_input.TryRead(out var result))
```

Comment:
How does this compare to calling PipeReader.ReadAsync in a non-async method and checking if it's completed? This would save the creation of a ValueTask, but in the sync case that ValueTask wouldn't be heap-allocated anyway, right?

Reply:
Would complicate the code as would have to pass the ValueTask to the IAsyncEnumerator; mark whether it had been `await`ed and skip it next time around the loop. 

The `while` loop in ReadAsync is the complicating factor

Reply:
Could look at `ManualResetValueTaskSourceCore` now have the shape of it down; but that would mean the types couldn't be shared as both read loops are different